### PR TITLE
HM6 I added login page, all concepts of login page and also I link DB…

### PR DESCRIPTION
**Pull request by AkhrorbekAbdurakhimov**
… from localhost

### CODE REVIEW

**Comment by vokanred** [comment]
Pls add me to reviewers and request review

File: client/src/app/auth/auth.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import { AuthService } from './services/auth.service';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+  constructor(private authservice: AuthService) {}
+
+  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
+    let token = localStorage.getItem('token')
+    if (this.authservice.isLoggedIn() && token) {
```
Comment (vokanred) [inline client/src/app/auth/auth.interceptor.ts]: if `this.authservice.isLoggedIn ` is true it means that token exists, so you don't need to check it again

File: client/src/app/auth/auth.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import { AuthService } from './services/auth.service';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+  constructor(private authservice: AuthService) {}
+
+  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
+    let token = localStorage.getItem('token')
```
Comment (vokanred) [inline client/src/app/auth/auth.interceptor.ts]: pls encapsulate this logic in auth service. create a method getToken() there.
auth interceptor shouldn't know about how and where you store the token

File: client/src/app/auth/login-form/login-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    email: ['', [Validators.required, Validators.email, Validators.pattern("^[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,4}$")]],
+    password: ['', [Validators.required]]
+  })
+
+  constructor(private fb: FormBuilder, private authService: AuthService, private router: Router) { }
+
+  onLogin() {
+    if (this.loginForm.valid) {
+      const { email, password } = this.loginForm.value
+      this.authService.login(email, password).subscribe({
```
Comment (vokanred) [inline client/src/app/auth/login-form/login-form.component.ts:27]: unsubscribe on destroy

File: client/src/app/config/app.config.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export default {
```
Comment (vokanred) [inline client/src/app/config/app.config.ts]: it's great that you decided to create a config but angular provides a solution for that
pls read https://angular.io/guide/build
Configuring application environments

File: client/src/app/auth/services/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  providedIn: 'root'
+})
+
+export class AuthService {
+
+  constructor(private http: HttpClient) { }
+
+  login(email: string, password: string) {
+    return this.http
+      .post<any>(`${config.APP.BASE_URL}/api/auth/login`, {
```
Comment (vokanred) [inline client/src/app/auth/services/auth.service.ts]: pls use environments

File: client/src/app/auth/auth.module.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+@NgModule({
+  declarations: [
+    LoginFormComponent
+  ],
+  imports: [
+    CommonModule,
+    SharedModule,
+    ReactiveFormsModule,
+    RouterModule.forChild(routes)
```
Comment (vokanred) [inline client/src/app/auth/auth.module.ts]: usually for child is used for lazy loaded modules

File: client/src/app/auth/auth.module.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import { CommonModule } from '@angular/common';
+import { ReactiveFormsModule } from '@angular/forms';
+import { RouterModule, Routes } from '@angular/router';
+
+import { SharedModule } from '../shared/shared.module';
+
+import { LoginFormComponent } from './login-form/login-form.component';
+
+const routes: Routes = [
+  { path: 'login', component: LoginFormComponent }
```
Comment (vokanred) [inline client/src/app/auth/auth.module.ts]: i think this route should be moved to app routing module

**Review by vokanred — CHANGES_REQUESTED** [review]
Pls review comments

File: client/src/app/config/app.config.ts (hunk 4 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/config/app.config.ts]: I create environments folder in src folder, I understand the environment.ts file will be replaced by environments.prod.ts file in the building

File: client/src/app/auth/auth.interceptor.ts (hunk 1 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/auth.interceptor.ts]: I removed token from checking the token exists

File: client/src/app/auth/auth.interceptor.ts (hunk 2 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/auth.interceptor.ts]: I created getToken() function inside of auth service

File: client/src/app/auth/services/auth.service.ts (hunk 5 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/services/auth.service.ts]: I used environments

File: client/src/app/auth/auth.module.ts (hunk 6 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/auth.module.ts]: I moved this logic to app routing module

File: client/src/app/auth/auth.module.ts (hunk 7 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/auth.module.ts]: I moved this logic to app routing module and I removed it from aut.module.ts

File: client/src/app/auth/login-form/login-form.component.ts (hunk 3 above)
Comment (AkhrorbekAbdurakhimov) [reply client/src/app/auth/login-form/login-form.component.ts:27]: ok, I will learn subscribe and destroy, then I will apply

**Review by vokanred — APPROVED** [review]
Done.

**Pull request merged**